Google.Api.Gax.Grpc/HeaderParameterExtractor.cs

```diff
@@ -0,0 +1,86 @@
+﻿/*
+ * Copyright 2021 Google Inc. All Rights Reserved.
+ * Use of this source code is governed by a BSD-style
+ * license that can be found in the LICENSE file or at
+ * https://developers.google.com/open-source/licenses/bsd
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text.RegularExpressions;
+
+namespace Google.Api.Gax.Grpc
+{
+    /// <summary>
+    /// 
+    /// </summary>
+    public sealed class HeaderParameterExtractor<TRequest>
+    {
+        private readonly IReadOnlyList<ParamExtraction> _extractions;
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="paramName"></param>
+        /// <param name="regex"></param>
+        /// <param name="selector"></param>
+        public HeaderParameterExtractor(string paramName, Regex regex, Func<TRequest, string> selector)
+        {
+            _extractions = new List<ParamExtraction> { new ParamExtraction(paramName, regex, selector) };
```

Comment:
All the tests show the parameter name being the same as the group name. If that's what we expect to do, let's build that in and simplify things:

- Call regex.GetGroupNames()
- Assert that the result has a single value
- Use the value as the parameter name

Reply:
👍 done